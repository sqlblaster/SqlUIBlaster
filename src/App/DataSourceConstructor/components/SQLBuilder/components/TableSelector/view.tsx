import { MenuItem, Select, Typography } from '@material-ui/core';
import * as React from 'react';
import { TableSelectorViewProps } from './props';

export const tablePickerPlaceholder = 'Choose a type of record';

export const TableSelector: React.FC<TableSelectorViewProps> = ({
  selectedTable,
  tables,
  handleTableSelection
}) => {
  return (
    <Select
      value={selectedTable ? selectedTable.modelName : tablePickerPlaceholder}
      onChange={handleTableSelection}
    >
      {!selectedTable && (
        <MenuItem
          disabled
          key={tablePickerPlaceholder}
          value={tablePickerPlaceholder}
        >
          <Typography variant='subtitle1'>{tablePickerPlaceholder}</Typography>
        </MenuItem>
      )}
      {tables &&
        tables.map(({ modelName }) => (
          <MenuItem button={true} value={modelName} key={modelName}>
            <Typography variant='subtitle1'>{modelName}</Typography>
          </MenuItem>
        ))}
    </Select>
  );
};
