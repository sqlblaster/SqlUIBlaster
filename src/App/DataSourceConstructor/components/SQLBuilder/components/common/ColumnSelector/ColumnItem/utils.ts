import { clone } from 'ramda';
import { ColumnBranch } from './SelectedColumn.models';

export const getNewTableBranch = (
  parentColumnBranch: ColumnBranch,
  foreignTableName: ColumnBranch['foreignTableName'],
  foreignModelName: ColumnBranch['foreignModelName']
) => {
  const newTableBranch = clone(parentColumnBranch);

  newTableBranch.lastColumn.foreignColumn = new ColumnBranch({
    tableName: foreignTableName,
    modelName: foreignModelName
  });

  newTableBranch.lastColumn = newTableBranch.lastColumn.foreignColumn;

  return newTableBranch;
};

export const listItemOwnPaddingLeft: number = 16;
export const paddingShiftAmountByLevel: number = 15;
export const calcPaddingAccordingTo = (level: number) => {
  return listItemOwnPaddingLeft + level * paddingShiftAmountByLevel;
};
