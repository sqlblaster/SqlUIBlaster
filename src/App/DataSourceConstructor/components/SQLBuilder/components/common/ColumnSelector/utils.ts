import { clone } from 'ramda';
import { Column } from 'src/App/DataSourceConstructor/schemas';
import { ColumnBranch } from './ColumnItem/SelectedColumn.models';
import { ColumnSelectorOwnProps } from './props';

export const getRootTableBranch = (tableName: string, modelName: string) =>
  new ColumnBranch({ tableName, modelName });

export const getNewColumnBranch = (
  tableBranch: ColumnBranch,
  { name, type, foreignTableName, foreignModelName }: Column
) => {
  const newColumnBranch = new ColumnBranch(clone(tableBranch));

  const lastColumn = newColumnBranch.lastColumn;
  lastColumn.columnName = name;
  lastColumn.type = type;
  lastColumn.foreignTableName = foreignTableName;
  lastColumn.foreignModelName = foreignModelName;

  return newColumnBranch;
};

export const sortInOrderForeignColumnsBelow = (columns: Column[]) =>
  columns.sort((c1, c2) => +!!c1.foreignTableName - +!!c2.foreignTableName);

export const filterColumns = (
  columns: Column[],
  canShowColumnBranch: Required<ColumnSelectorOwnProps>['canShowColumnBranch'],
  tableBranch: ColumnBranch
): Column[] => {
  return columns.filter(
    column =>
      canShowColumnBranch(getNewColumnBranch(tableBranch, column)) ||
      column.foreignTableName
  );
};
