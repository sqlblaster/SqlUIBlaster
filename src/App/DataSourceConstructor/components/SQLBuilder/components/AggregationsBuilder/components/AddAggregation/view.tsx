import * as React from 'react';
import { AddEntity } from '../../../common/AddEntity';
import { MutateAggregationPopover } from '../MutateAggregationPopover';
import { AddAggregationViewProps } from './props';

export const addAggregationText = 'Add an aggregation';

// FIXME: make padding clickable too
export class AddAggregation extends React.Component<AddAggregationViewProps, any> {
  private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

  constructor(props: AddAggregationViewProps) {
    super(props);

    this.popoverLauncher = React.createRef<HTMLDivElement>();
  }

  public render() {
    const {
      aggregations,
      addAggregation,
      isOpen,
      handleClose,
      handleOpen
    } = this.props;

    return (
      <>
        <AddEntity
          showText={!(aggregations && aggregations.length)}
          onClick={handleOpen}
          reference={this.popoverLauncher}
        >
          {addAggregationText}
        </AddEntity>
        { isOpen && (// TODO: resolve this problem another way
          <MutateAggregationPopover
            action={addAggregation}
            isOpen={isOpen}
            anchorEl={this.popoverLauncher.current}
            onClose={handleClose}
          />
        ) }
      </>
    );
  }
}
