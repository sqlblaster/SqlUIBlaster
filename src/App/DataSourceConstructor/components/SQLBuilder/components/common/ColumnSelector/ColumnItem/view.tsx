import {
  Collapse,
  IconButton,
  ListItem,
  ListItemIcon,
  ListItemText,
  withStyles
} from '@material-ui/core';
import ExpandLess from '@material-ui/icons/ExpandLess';
import ExpandMore from '@material-ui/icons/ExpandMore';
import * as React from 'react';
import { TableColumnsDelegatedProps } from '../props';
import { TableColumns } from '../view';
import { ColumnItemViewProps } from './props';
import { TypeIcon } from './TypeIcon';
import { calcPaddingAccordingTo, getNewTableBranch } from './utils';

export const ColumnItem = withStyles({})(
  class extends React.Component<ColumnItemViewProps> {
    public render() {
      const {
        name,
        type,
        foreignTableName,
        foreignModelName,
        columnBranch,
        level,
        tables,
        popoverPositionUpdater,
        onColumnSelected,
        canShowColumnBranch,
        collapsed,
        handleColumnClick,
        handleCollapsedStateChange,
        handleCollapseEnter,
        handleCollapseExit
      } = this.props;

      const delegatedProps: TableColumnsDelegatedProps = {
        tables,
        popoverPositionUpdater,
        onColumnSelected,
        canShowColumnBranch
      };

      const disabledForeignColumn = canShowColumnBranch
        ? !canShowColumnBranch(columnBranch) && !!foreignTableName
        : false;

      return (
        <>
          <ListItem
            {...(disabledForeignColumn
              ? {}
              : { button: true, onClick: handleColumnClick })}
            style={{
              paddingLeft: calcPaddingAccordingTo(level),
              ...(disabledForeignColumn ? { cursor: 'default' } : {})
            }}
          >
            <ListItemIcon>
              <TypeIcon type={foreignTableName ? 'foreign-column' : type} />
            </ListItemIcon>
            <ListItemText inset={true} primary={name} />
            {foreignTableName && (
              <>
                {collapsed ? (
                  <IconButton
                    onClick={handleCollapsedStateChange}
                    style={{ padding: 0 }}
                  >
                    <ExpandMore />
                  </IconButton>
                ) : (
                  <IconButton
                    onClick={handleCollapsedStateChange}
                    style={{ padding: 0 }}
                  >
                    <ExpandLess />
                  </IconButton>
                )}
              </>
            )}
          </ListItem>

          {foreignTableName && (
            <Collapse
              in={!collapsed}
              timeout='auto'
              unmountOnExit={true}
              onEntered={handleCollapseEnter}
              onExited={handleCollapseExit}
            >
              <TableColumns
                {...delegatedProps}
                level={level + 1}
                tableName={foreignTableName}
                tableBranch={getNewTableBranch(
                  columnBranch,
                  foreignTableName,
                  foreignModelName
                )}
              />
            </Collapse>
          )}
        </>
      );
    }
  }
);
