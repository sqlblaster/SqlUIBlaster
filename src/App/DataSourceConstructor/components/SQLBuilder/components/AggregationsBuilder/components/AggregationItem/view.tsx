import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { ColumnBranch } from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/common/ColumnSelector/ColumnItem/SelectedColumn.models';
import { RemoveButton } from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/RemoveButton';
import { SequentialColumnView } from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/common/SequentialColumnView';
import { MutateAggregationPopover } from '../MutateAggregationPopover';
import { AggregationItemViewProps } from './props';
import { aggregationItemStyles } from './styles';

export const AggregationItem = withStyles(aggregationItemStyles)(
  class extends React.Component<AggregationItemViewProps> {
    private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

    constructor(props: AggregationItemViewProps) {
      super(props);
      this.popoverLauncher = React.createRef<HTMLDivElement>();
    }

    public render() {
      const {
        aggregation,
        updateAggregation,
        classes,
        isOpen,
        handleOpen,
        handleClose,
        handleAggregationRemoval
      } = this.props;

      const { id, operator, column } = aggregation;

      return (
        <>
          <div className={classes.root}>
            <div
              ref={this.popoverLauncher}
              onClick={handleOpen}
              className={classes['column-view']}
            >
              <SequentialColumnView
                prefix={operator}
                {...(column ? column : new ColumnBranch())}
              />
            </div>
            <RemoveButton onClick={handleAggregationRemoval(id)} />
          </div>
          { isOpen && (
            <MutateAggregationPopover
              isOpen={isOpen}
              anchorEl={this.popoverLauncher.current}
              action={updateAggregation}
              onClose={handleClose}
              aggregation={aggregation}
            />
          ) }
        </>
      );
    }
  }
);
