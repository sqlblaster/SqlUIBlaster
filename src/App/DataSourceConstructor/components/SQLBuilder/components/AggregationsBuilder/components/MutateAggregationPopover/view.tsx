import { List, ListItem, ListItemText, withStyles } from '@material-ui/core';
import * as React from 'react';
import { ColumnSelector } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/ColumnSelector';
import { BackButton } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/BackButton';
import { PopoverPaper } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/PopoverPaper';
import { aggregationOperators } from '../../aggregations';
import { Aggregation } from '../AggregationItem/model';
import { AggregationTips } from './AggregationTips';
import { MutateAggregationPopoverViewProps } from './props';
import { mutateAggregationPopoverStyles } from './styles';

export const MutateAggregationPopover = withStyles(
  mutateAggregationPopoverStyles
)((props => {
  const {
    column,
    operator,
    aggregation,
    popoverPositionUpdater,
    classes,
    onColumnSelected,
    onOperatorSelected,
    onBackClicked,
    canShowColumnBranch
  } = props;

  return (
    <PopoverPaper {...props}>
      <div className={classes.root}>
        {((aggregation &&
          Aggregation.hasNoColumn(operator || aggregation.operator)) ||
          !operator) && (
          <div className={classes['select-operator']}>
            <List component='nav' disablePadding={true}>
              {aggregationOperators.map(aggregationOperator => (
                <ListItem
                  key={aggregationOperator}
                  button={true}
                  selected={aggregationOperator === operator}
                  onClick={onOperatorSelected(aggregationOperator)}
                >
                  <ListItemText
                    inset={true}
                    primary={aggregationOperator}
                    style={{ padding: 0 }}
                  />
                  <AggregationTips operator={aggregationOperator} />
                </ListItem>
              ))}
            </List>
          </div>
        )}

        {operator &&
          !Aggregation.hasNoColumn(operator) &&
          ((aggregation && operator && column) || (operator && !column)) && (
            <>
              <div className={classes['title-component']}>
                <BackButton onClick={onBackClicked} />
                {`${operator}...`}
              </div>
              <hr className={classes.hr} />
              <ColumnSelector
                popoverPositionUpdater={popoverPositionUpdater}
                canShowColumnBranch={canShowColumnBranch}
                onColumnSelected={onColumnSelected}
              />
            </>
          )}
      </div>
    </PopoverPaper>
  );
}) as React.FC<MutateAggregationPopoverViewProps>);
