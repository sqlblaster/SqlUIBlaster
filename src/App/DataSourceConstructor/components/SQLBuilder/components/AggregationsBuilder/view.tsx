import { Typography, withStyles } from '@material-ui/core';
import * as React from 'react';
import { AddAggregation } from './components/AddAggregation';
import { AggregationItem } from './components/AggregationItem';
import { AggregationsBuilderViewProps } from './props';
import { aggregationsBuilderStyles } from './styles';

export const AggregationsBuilder = withStyles(aggregationsBuilderStyles)((({
  aggregations,
  classes
}) => (
  <div className={classes.root}>
    {aggregations && aggregations.length ? (
      <div className={classes.aggregations}>
        {aggregations.map((aggregation, index) => (
          <React.Fragment key={aggregation.id}>
            {index !== 0 && (
              <Typography variant='caption' className={classes['and-operator']}>
                and
              </Typography>
            )}
            <AggregationItem aggregation={aggregation} />
          </React.Fragment>
        ))}
      </div>
    ) : null}
    <AddAggregation />
  </div>
)) as React.FC<AggregationsBuilderViewProps>);
