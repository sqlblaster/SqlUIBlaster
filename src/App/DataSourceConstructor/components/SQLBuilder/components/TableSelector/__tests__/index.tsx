import { getGeneratedQuery, render } from '../../../__tests__/render.utils';
import { selectATable } from '../../TableSelector/__tests__/utils/select-table';
import { tablePickerPlaceholder } from '../view';

describe('TableSelector', () => {
  const { componentNode } = render();

  it('should show placeholder by default', () => {
    expect(componentNode).toHaveTextContent(new RegExp(tablePickerPlaceholder));
  });

  it('should change the selected table when one is picked', async () => {
    selectATable();

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it.skip('should clear all query entities including drill-down entities', () => {});
});
