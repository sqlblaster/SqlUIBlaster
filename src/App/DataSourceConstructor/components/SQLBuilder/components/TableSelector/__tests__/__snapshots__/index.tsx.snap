// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSelector should change the selected table when one is picked 1`] = `
"SELECT
  *
FROM
  \\"orders\\"
LIMIT 2000
"
`;
